### DISCUSSION TITLE
Bulk import from directory of JSON Schema files

### DISCUSSION THREAD

**Discussion by jdpedrie** [TandoorRecipes/recipes#2729] [Q&A] [answered]
I have several hundred recipes stored in individual JSON Schema files. A manual import of each works just fine, but this is pretty tedious work. Is there a way to import them all at once? An API would be just fine.

**Comment by smilerz** [comment, marked as answer]
there is a fully functional API - documented at /docs/api/

**Reply by jdpedrie** [reply]
Sorry, I'm using `/api/recipe-from-source/`, and got the authentication working fine, but keep getting 400 errors,

```json
{
	"data": "{\"$context\": \"http://schema.org\"}",
    "mode": "source",
    "auto": "true"
}
```

This returns "Invalid data. Expected a dictionary, but got str."

I've tried varying requests, from setting `data` to the un-escaped JSON, to omitting mode and auto, to nesting the entire document in a `data` property.

**Reply by smilerz** [reply]
can you enable debug and provide the full trace

**Reply by jdpedrie** [reply]
Running in docker. I enabled debug, and see DEBUG=1 in the environment when I inspect the container. Also set GUNICORN_LOG_LEVEL to debug. Not getting a trace, just the following.

```
Bad Request: /api/recipe-from-source/
Bad Request: /api/recipe-from-source/
172.29.8.4 - - [10/Nov/2023:12:53:45 -0500] "POST /api/recipe-from-source/ HTTP/1.0" 400 74 "-" "axios/1.5.1"
[2023-11-10 12:53:45 -0500] [13] [DEBUG] Closing connection.
```

**Reply by smilerz** [reply]
what version are you running?  that api needs two fields: data and url
data can either be a stringified (I think?) json doc or an html doc

**Reply by jdpedrie** [reply]
I was running the `vabene1111/recipes:latest`, but I just tried your `smilerz/recipes:latest` as well, and got the same thing. When I send the document as JSON (i.e. not as an escaped string), I get "Not a valid string."

Looking at the [handler](https://github.com/TandoorRecipes/recipes/blob/8a588db4296f58ddbfa2cf8c5e4d9697c1fb9a6c/cookbook/views/api.py#L1281), it seems that data and URL don't both need to be specified. If URL is omitted, it will try to parse data as a JSON schema.

That said, I'm fairly certain it's failing earlier, with `serializer.is_valid()` returning false, though it's a bit hard to be certain.

**Reply by smilerz** [reply]
right - the json needs to be stringfied

Can you produce a minimal example of the complete payload you are sending to the api endpoint?

I suspect one of the fields provided is being passed as a string instead of a dictionary, but I'd need to see an example of what you are producing.
